Sources/Tools/plFilePatcher/main.cpp

```diff
@@ -0,0 +1,186 @@
+        GetConsoleScreenBufferInfo(GetStdHandle(STD_OUTPUT_HANDLE), &csbi);
+        width = size_t(csbi.srWindow.Right - csbi.srWindow.Left);
+    }
+#else
+#   include <sys/ioctl.h>
+#   include <unistd.h>
+
+    void GetConsoleWidth(size_t& width)
+    {
+        struct winsize w;
+        ioctl(fileno(stdout), TIOCGWINSZ, &w);
+        width = size_t(w.ws_col) - 1;
+    }
+#endif
+
+void PrintUsage(const ST::string& exeName)
+{
+    ST::printf("Plasma File Patcher\n{}\n\n", plProduct::ProductString());
+
+    ST::printf("File patcher for the Plasma client and game data.\n\n");
+
+    ST::printf("Usage:\n");
+    ST::printf("\t{} [options]\n\n", exeName);
+
+    ST::printf("Available options:\n");
+    ST::printf("\t--ServerIni file.ini\tSpecify the server config file\n");
+    ST::printf("\t--DataOnly\t\tPatch only the game data files\n");
+    ST::printf("\t--ClientOnly\t\tPatch only the client executable files\n");
+    ST::printf("\t--help\t\t\tDisplay this help message\n");
+    ST::printf("\t--quiet\t\t\tDo not display download progress\n\n");
+
+    ST::printf("By default, updates to both game data and client executables will be\ndownloaded, and it will use \"server.ini\" for server connection details.\n");
+}
+
+size_t consoleWidth = 0;
+ST::string lastProgress;
+
+void FileDownloadBegin(const plFileName& file)
+{
+    // To account for terminal resizing, we want to refresh this value
+    // periodically, but not every time we draw to the screen, so let's grab it
+    // whenever we start downloading a new file (which is probably still
+    // excessive tbh)
+    GetConsoleWidth(consoleWidth);
+
+    size_t fill = consoleWidth - file.GetSize();
+    ST::printf("\r{}{}\n{}", file, ST::string::fill(fill, ' '), lastProgress);
+}
+
+void FileDownloadProgress(uint64_t bytesDown, uint64_t bytesTotal, const ST::string& stats)
+{
+    size_t availWidth = consoleWidth - stats.size() - 3;
+    size_t progWidth = size_t(double(bytesDown) / double(bytesTotal) * availWidth);
+
+    lastProgress = ST::format("|{}{}| {}", ST::string::fill(progWidth, '='), ST::string::fill(availWidth - progWidth, ' '), stats);
+    ST::printf("\r{}", lastProgress);
+}
+
+int main(int argc, char* argv[])
+{
+    enum { kArgServerIni, kArgDataOnly, kArgClientOnly, kArgQuiet, kArgHelp1, kArgHelp2 };
+    const plCmdArgDef cmdLineArgs[] = {
+        { kCmdArgFlagged | kCmdTypeString,  "ServerIni",    kArgServerIni },
+        { kCmdArgFlagged | kCmdTypeBool,    "DataOnly",     kArgDataOnly },
+        { kCmdArgFlagged | kCmdTypeBool,    "ClientOnly",   kArgClientOnly },
+        { kCmdArgFlagged | kCmdTypeBool,    "quiet",        kArgQuiet },
+        { kCmdArgFlagged | kCmdTypeBool,    "help",         kArgHelp1 },
+        { kCmdArgFlagged | kCmdTypeBool,    "?",            kArgHelp2 },
+    };
+
+    std::vector<ST::string> args;
+    args.reserve(argc);
+    for (size_t i = 0; i < argc; i++) {
+        args.emplace_back(ST::string::from_utf8(argv[i]));
```

Comment:
On Windows, `argv` will be in the ACP (generally latin_1 for our audience).

Reply:
This matches what plUruLauncher is doing 🤷🏼‍♂️